### ISSUE TITLE
Why don't ML engineers use shampoo ?🧴

### ISSUE THREAD

**Issue by G-structure** [huggingface/accelerate#3178]
Hey,

I have been using Meta's Implementation of Distributed Shampoo and am seeing ~20% faster convergence of transformer based models compared to AdamW. [Simo Ryu](https://x.com/cloneofsimo/status/1836003682141577418) has done some nice investigations into the advantages of Shampoo. 

I am looking to use [Shampoo](https://github.com/facebookresearch/optimizers/tree/main/distributed_shampoo) and [Soap](https://github.com/nikhilvyas/SOAP) as an optimizer in accelerate but their current implementations introduce some breaking changes.

Focusing on Shampoo for now:

Distributed Shampoo disabled `state_dict` and `load_state_dict` in favor of a custom `distributed_state_dict `, `load_distributed_state_dict`. Both of which require the models `named_parameters()` to be passed in as args. More info as to why [here](https://github.com/facebookresearch/optimizers/blob/main/distributed_shampoo/README.md#checkpointing-support)

I have a [hacky commit here](https://github.com/huggingface/accelerate/commit/6aa5176c76d57203880bfe47cec0c5ae712fea43#diff-5fc6d57c56dbd4a9c5518cf23576a66fb4aba9c5222d90c4e62ec6db083dca79R120) to patch `accelerate/optimizers`. However I am still forced to bypass `accelerate.save()` and use `dist_checkpoint.save_state_dict()` directly since the optimizer in the state_dict needs to have access to the models `named_parameters()`. 
```
state_dict = {
    "model": model.state_dict(),
    "optim": optimizer.distributed_state_dict(key_to_param=model.named_parameters()),
}
dist_checkpoint.save_state_dict(
    state_dict=state_dict,
    storage_writer=dist_checkpoint.FileSystemWriter(CHECKPOINT_DIR),
)
```
You can see this here in my [e2-tts training code](https://github.com/cyber-phys/e2-tts-shampoo/blob/aaf6a4f1b9fe7a7aa7112b4c4f86f7a3bc05eff7/e2_tts_pytorch/trainer.py#L217). I am able to save the model weights but am not yet able to load them again when using accelerate. This is where I am lost currently. 

Also since I don't have access to the `named_parameters` until `accelerate.prepare_model()` is called the shampoo optimizer needs to be defined in the model definition, which makes it awkward to switch between optimizers, [see here](https://github.com/cyber-phys/e2-tts-shampoo/blob/aaf6a4f1b9fe7a7aa7112b4c4f86f7a3bc05eff7/e2_tts_pytorch/trainer.py#L182)

Ideally id be able to do something like [this](https://github.com/cyber-phys/e2-tts-shampoo/blob/aaf6a4f1b9fe7a7aa7112b4c4f86f7a3bc05eff7/experiments/train_with_shampoo.py) where I pass in the optimizer as I can with AdamW.
```
e2tts = E2TTS(
    cond_drop_prob=0.0,
    transformer = dict(
        dim = 512,
        depth = 2,
        heads = 6,
        skip_connect_type = 'concat'
    ),
    mel_spec_kwargs = dict(
        filter_length = 1024,
        hop_length = 256,
        win_length = 1024,
        n_mel_channels = 100,
        sampling_rate = 24000,
    ),
    frac_lengths_mask = (0.7, 0.9)
)


optimizer = DistributedShampoo(
    e2tts.parameters(),
    lr=7.5e-5,
    betas=(0.9, 0.999),
    epsilon=1e-12,
    weight_decay=1e-05,
    max_preconditioner_dim=8192,
    precondition_frequency=100,
    use_decoupled_weight_decay=False,
        grafting_config=AdamGraftingConfig(
            beta2=0.999,
            epsilon=1e-08,
        ),
    )

trainer = E2Trainer(
    e2tts,
    optimizer
)

trainer.train(train_dataset, epochs, batch_size, save_step=50)
```

ofc when I setup everything with torch ddp, instead of accelerate everything works as intended :/

What would be the best approach for accelerate to support these custom optimizers (ones not part of torch)? My plan currently is to write a ShampooPlugin along the lines of the DeepSpeedPlugin, but it would be nice if the shampoo optimizer could be detected automatically without having to change the accelerate config. I am willing to put in the work to solve this so more projects can benefit from using these new optimizers with accelerate.

Any guidance would be much appreciated.  :)

**Comment by bghira** [comment]
that optim lacks other torch-specific expectations and like learning rate schedulers, some off-the-wall optimisers just don't work without modification. that distributed zero shampoo optim is a WIP technical prototype and not meant to be used in production, for example.

the SOAP one worked as expected for me. it just needs Closure input on its step. see here: https://github.com/bghira/SimpleTuner/blob/main/helpers/training/optimizers/soap/__init__.py

**Comment by bghira** [comment]
other problems of the original optim implementations linked is that they are not functioning with torch.compile and retain very slow performance (exaggerated in SOAP) and high memory overhead (also exaggerated in SOAP) even with ZeRO offload
